h3 3.6.4 OSX builds failed

github action log, https://github.com/Homebrew/homebrew-core/runs/792748977

<details>
<summary>build failure log</summary>

```
[  8%] Linking C shared library lib/libh3.dylib
/usr/local/Cellar/cmake/3.17.3/bin/cmake -E cmake_link_script CMakeFiles/h3WithTestAllocators.dir/link.txt --verbose=1
/usr/local/Cellar/cmake/3.17.3/bin/cmake -E cmake_link_script CMakeFiles/h3.dir/link.txt --verbose=1
/usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -dynamiclib -Wl,-headerpad_max_install_names -compatibility_version 1.0.0 -o lib/libh3.1.dylib -install_name @rpath/libh3.1.dylib CMakeFiles/h3.dir/src/h3lib/lib/algos.c.o CMakeFiles/h3.dir/src/h3lib/lib/coordijk.c.o CMakeFiles/h3.dir/src/h3lib/lib/bbox.c.o CMakeFiles/h3.dir/src/h3lib/lib/polygon.c.o CMakeFiles/h3.dir/src/h3lib/lib/h3Index.c.o CMakeFiles/h3.dir/src/h3lib/lib/vec2d.c.o CMakeFiles/h3.dir/src/h3lib/lib/vec3d.c.o CMakeFiles/h3.dir/src/h3lib/lib/linkedGeo.c.o CMakeFiles/h3.dir/src/h3lib/lib/localij.c.o CMakeFiles/h3.dir/src/h3lib/lib/geoCoord.c.o CMakeFiles/h3.dir/src/h3lib/lib/h3UniEdge.c.o CMakeFiles/h3.dir/src/h3lib/lib/mathExtensions.c.o CMakeFiles/h3.dir/src/h3lib/lib/vertexGraph.c.o CMakeFiles/h3.dir/src/h3lib/lib/faceijk.c.o CMakeFiles/h3.dir/src/h3lib/lib/baseCells.c.o  /usr/lib/libm.dylib 
/usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -dynamiclib -Wl,-headerpad_max_install_names -compatibility_version 1.0.0 -o lib/libh3WithTestAllocators.1.dylib -install_name @rpath/libh3WithTestAllocators.1.dylib CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/algos.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/coordijk.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/bbox.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/polygon.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/h3Index.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/vec2d.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/vec3d.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/linkedGeo.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/localij.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/geoCoord.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/h3UniEdge.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/mathExtensions.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/vertexGraph.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/faceijk.c.o CMakeFiles/h3WithTestAllocators.dir/src/h3lib/lib/baseCells.c.o  /usr/lib/libm.dylib 
Undefined symbols for architecture x86_64:
  "_test_prefix_calloc", referenced from:
      _kRingDistances in algos.c.o
      __polyfillInternal in algos.c.o
      _compact in h3Index.c.o
      _addNewLinkedPolygon in linkedGeo.c.o
      _addNewLinkedLoop in linkedGeo.c.o
      _normalizeMultiPolygon in linkedGeo.c.o
      _initVertexGraph in vertexGraph.c.o
      ...
  "_test_prefix_free", referenced from:
      _kRingDistances in algos.c.o
      __polyfillInternal in algos.c.o
      _compact in h3Index.c.o
      _destroyLinkedGeoLoop in linkedGeo.c.o
      _destroyLinkedPolygon in linkedGeo.c.o
      _normalizeMultiPolygon in linkedGeo.c.o
      _destroyVertexGraph in vertexGraph.c.o
      ...
  "_test_prefix_malloc", referenced from:
      __polyfillInternal in algos.c.o
      _compact in h3Index.c.o
      _addLinkedCoord in linkedGeo.c.o
      _normalizeMultiPolygon in linkedGeo.c.o
      _addVertexNode in vertexGraph.c.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
/usr/local/Cellar/cmake/3.17.3/bin/cmake -E cmake_symlink_library lib/libh3.1.dylib lib/libh3.1.dylib lib/libh3.dylib
make[2]: *** [lib/libh3WithTestAllocators.1.dylib] Error 1
make[1]: *** [CMakeFiles/h3WithTestAllocators.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[  8%] Built target h3
make: *** [all] Error 2
```

</details>

PR reference, https://github.com/Homebrew/homebrew-core/pull/56633

cc @robsalasco

This is an older post but it looks like this may have to do with XCode defaulting to not looking up undefined symbols dynamically: https://glandium.org/blog/?p=2764 I was able to reproduce this problem with XCode 11.5 installed. 

The undefined symbols are only used in tests to confirm that the library can built with undefined symbols for memory allocation that are then provided at runtime. If that functionality isn't being used, we could just disable the test.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,8 @@ The public API of this library consists of the functions declared in file
 [h3api.h.in](./src/h3lib/include/h3api.h.in).
 
 ## [Unreleased]
+### Fixed
+- Fixed building the library with custom memory allocation functions on Mac OSX. (#356)
 
 ## [3.6.4] - 2020-06-19
 ### Added

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -218,11 +218,20 @@ function(add_h3_library name h3_alloc_prefix_override)
     endif()
 
     target_compile_definitions(${name} PUBLIC H3_PREFIX=${H3_PREFIX})
+    set(has_alloc_prefix NO)
     if(h3_alloc_prefix_override)
+        set(has_alloc_prefix YES)
         target_compile_definitions(${name} PUBLIC H3_ALLOC_PREFIX=${h3_alloc_prefix_override})
     elseif(H3_ALLOC_PREFIX)
+        set(has_alloc_prefix YES)
         target_compile_definitions(${name} PUBLIC H3_ALLOC_PREFIX=${H3_ALLOC_PREFIX})
     endif()
+    # Mac OSX defaults to not looking up undefined symbols dynamically,
+    # so enable that explicitly.
+    if(has_alloc_prefix AND APPLE)
+        target_link_libraries(${name} PRIVATE "-undefined dynamic_lookup")
+    endif()
+
     if(have_alloca)
         target_compile_definitions(${name} PUBLIC H3_HAVE_ALLOCA)
     endif()

